In [1]:
from datascience import *
import numpy as np
from math import *
from scipy import stats
import matplotlib.pyplot as plt
%matplotlib inline

## Hypothesis Testing

In Data8.2x, you have been using simulation to conduct hypothesis testing. Now that we have completed Data8.2x, this is a good time to take a step back and reflect on hypothesis testing. 

Every hypothesis test has roughly the same structure. The following 4 steps provide a rough outline:

1) State the null and alternate hypotheses. Generally, the alternate hypothesis is what you are trying to show. Essentially, if you want to show a result, we assume the opposite is true and try to prove ourselves wrong. 

2) Determine/calculate a test statistic. See your book for a formal definition, but generally, the test statistic is any quantity that helps us evaluate our sample with respect to our null hypothesis. 

3) Determine distribution of test statistic and compute a $p$-value. If you have taken inferential statistics before, you likely computed a $z$ or $t$ statistic and used a calculator or table to compute a $p$-value. This is based on asymptotic theory of sample means/proportions. This is not the approach taken in Data8.2x. With better computing power, we can use simulation to obtain an empirical distribution of our test statistic under the null hypothesis. 

4) Conclude. For low $p$-value (generally below 0.05), we reject the null hypothesis. For high $p$-value, we fail to reject. Low $p$-value implies that our sample would be very unusual if the null hypothesis were actually true. Therefore, that is evidence that the null hypothesis is wrong. 

### Example

Let's work through an example. Suppose that in the upcoming election, Referendum A is up for approval in Colorado. You suspect that in El Paso County, more than half of eligible voters support the referendum. You collect a random sample of 200 eligible voters in El Paso County and 115 of them express support. Is there evidence to conclude that supports your suspicion? 

#### Step 1: Hypothesis

State the null and alternative hypotheses.

null: H0: pi <= 0.5  
alternative: Ha: pi > 0.5

#### Step 2: Test Statistic

Select a test statistic and compute that test statistic for the sample.

potential test statistic options:  
x = # of people out of 200 who vote for referendum A  
x = # of people in excess of 100 out of 200 who vote in favor of referendum A  
p = # of people out of 200 who vote in favor of referendum A divided by 200 (gives a proprtion estimate for unknown population)  
  
using option 2 above for test statistic:  
for this sample x = 15

#### Step 3: $p$-value

3a) If $H_0$ were true, what should the value of $\hat{p}$ be close to? In other words, if in fact, half of eligible voters support the referendum, what value should your test statistic take? 

If H0 were true, the value of x should be close to zero.

3b) In words (and in the context of this problem), describe what the $p$-value is. 

If the proportion of voters in favor of the referndum were 0.5, the p-value is the propability of getting 15 or more people in favor and in excess of 100 out of 200.

3c) Find the $p$-value directly and using simulation. Hint: the binomial distribution will be of use here. 

In [2]:
#NOTE: I used several python boxes, so don't feel the need to put everything in this box.
my_binom = stats.binom(n=200,p=.5,loc=-100)

In [3]:
# probability calculation for our result
# P(>=15) = 1 - P(<=14)
1 - my_binom.cdf(14)

0.020018595806698514

#### Step 4: Conclude

What is your conclusion? Be sure to state your conclusion in the context of the problem.

Reject the null (because of the low p-value, p=.02 < .05 which is a common threshhold for what constitutes a low p-value). The low p-value implies that our sample would be hihgly unlikely if the null hypothesis were true. This is evidence that the null hypothesis is wrong and in support of our alternative hypothesis. 

## Confidence Intervals

Construct and interpret a 95% confidence interval on $p$, the true proportion of eligible El Paso County voters who support the referendum. There are many ways to construct such an interval (bootstrap, the binomial distribution, asymptotically). Select one and implement. 

Also, compare your interval to the results of your hypothesis test. Does your interval contain the value 0.5? Why does that matter? 

In [3]:
# rewrite data using zeros and ones to symbolize the two voting options
my_data = np.repeat([0,1],[85,115])
np.random.seed(951)
my_boot = []
for _ in np.arange(10000):
    my_boot = np.append(my_boot, sum(np.random.choice(my_data,size=200))-100)

In [4]:
test_stat = Table().with_column('test stat',my_boot)
test_stat.percentile(5)

test stat
3

In [5]:
103/200

0.515

We are 95% confident that the true proportion of voters in favor of referendum A is > 103/200 or 51.5%. The value .5 is not in the confidence interval which matters because it supports our earlier decision to reject the null.